packages/scrapers-v2/src/urls/scrape_urls.ts

```diff
@@ -0,0 +1,162 @@
+  }
+
+  if (start !== 2021) {
+    // todo: implement generic scraping from overall archive
+    // https://registrar.northeastern.edu/group/academic-catalogs/
+    throw new Error("only current year is supported");
+  }
+
+  return scrapeMajorLinksForUrl(
+    "https://catalog.northeastern.edu",
+    "undergraduate"
+  );
+};
+
+/**
+ * Given a baseUrl and a path, attempts to scrape the major catalog based on the sidebar
+ * hierarchy.
+ * @param baseUrl the base url of the major catalog. should look something
+ *                like "https://catalog.northeastern.edu"
+ * @param path    the path of the major catalog. something like "/undergraduate"
+ *                or "archive/2018-2019/undergraduate/". trailing or leading slashes are ok.
+ */
+export const scrapeMajorLinksForUrl = async (
+  baseUrl: string,
+  path: string
+): Promise<CatalogHierarchy> => {
+  const paths = getPathParts(path);
+  try {
+    const initStack = Object.values(College).map((college) => ({
+      path: [...paths, college],
+    }));
+    const catalogPaths = await scrapeLinks(baseUrl, initStack);
+    return convertToHierarchy(baseUrl, catalogPaths);
+  } catch (e) {
+    throw e;
+  }
+};
+
+/**
+ *
+ * Retrieves all sub-entries of the given initial queue in BFS fashion using the catalog sidebar hierarchy.
+ *  
+ * @param baseUrl   the base catalog URL, i.e. https://catalog.northeastern.edu
+ * @param initQueue a queue of parent entries
+ * @returns         a flat list of all the last level children catalog entries
+ */
+const scrapeLinks = async (
+  baseUrl: string,
+  initQueue: CatalogPath[]
+): Promise<CatalogPath[]> => {
+  const results: CatalogPath[] = [];
+
+  let queue = initQueue;
+  while (queue.length > 0) {
+    const pages = await Promise.all(getUrlHtmls(queue, baseUrl));
+    const nextQueue: CatalogPath[] = [];
+    for (const { $, url } of pages) {
+      const children = getChildrenForPathId($, url).toArray().map($);
+      for (const element of children) {
+        const path = getLinkForEl(element);
+        const bucket = isParent(element) ? nextQueue : results;
+        bucket.push({ path });
+      }
+    }
+    queue = nextQueue;
+  }
+
+  return results;
+};
+
+/**
+ * Converts a flat list of entries to catalog hierarchy.
+ * 
+ * @param base         the base catalog URL, i.e. https://catalog.northeastern.edu
+ * @param catalogPaths a flat list of paths
+ * @returns            catalog hierarchy
+ */
+const convertToHierarchy = (
+  base: string,
+  catalogPaths: CatalogPath[]
+): CatalogHierarchy => {
+  const hierarchy: CatalogHierarchy = {};
+  for (const { path } of catalogPaths) {
+    let obj: CatalogHierarchy = hierarchy;
+
+    // For each part of the path, add it to the hierarchy
+    for (let i = 0; i < path.length - 1; i += 1) {
+      const part = path[i];
+      if (!(part in obj)) {
+        obj[part] = {};
+      }
+      const child = obj[part];
+      if (typeof child === "string") {
```

Comment:
so for `CatalogHierarchy`, it is either `string | CatalogHierarchy`. Basically we only want to add to the object when it's a `CatalogHierarchy`

Reply:
yeah, this shouldn't ever happen. but here's an example of what would be considered "bad":

- a major exists at the path: `coe/marine-bio`
- another major exists at the path: `coe/marine-bio/fishing-concentration`

we parse the first one first, and add a leaf at `hierarchy[coe][marine-bio]`. then, we go to add the `fishing-concentration` one, but when we get `hierarchy[coe][marine-bio]`, we can't index it, because it's a string. oops!